### PR TITLE
Do transpose on concat's inputs instead of output

### PR DESCRIPTION
**Pull request by tungld**
This patch adds on canonicalization rule to ONNXTranposeOp which does tranpose on concat's inputs instead of output.

Signed-off-by: Tung D. Le <tung@jp.ibm.com>

### CODE REVIEW

File: src/Dialect/ONNX/Rewrite.td [hunk 1]
```diff
@@ -309,6 +313,28 @@ def RemoveIdentityTransposePattern:  Pat<
   // Check that we have indeed a identity transpose pattern.
   [(IsIdentityPermuteAttribute:$p)]>;
 
+def GetIndexOfAxisInPerm: NativeCodeCall<
+"[](PatternRewriter &rewriter, ArrayAttr permAttr, IntegerAttr axis) {"#
+"  IntegerAttr result;"#
+"  for (uint64_t i = 0; i < permAttr.getValue().size(); ++i) {"#
+"    if (permAttr.getValue()[i].cast<IntegerAttr>().getValue().getSExtValue()"#
```
Comment (etiotto) [inline src/Dialect/ONNX/Rewrite.td:320]: Should defensively check whether the cast is legal (assert)

File: src/Dialect/ONNX/Rewrite.cpp [hunk 2]
```diff
@@ -63,6 +63,20 @@ Type getReturnTypeForMatMulOpND2D(Value A, Value B) {
       resShape, A.getType().cast<ShapedType>().getElementType());
 }
 
+std::vector<Value> transposeVariadicInput(PatternRewriter &rewriter,
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/Dialect/ONNX/Rewrite.cpp]: Done. Thanks!

File: src/Dialect/ONNX/Rewrite.cpp [hunk 3]
```diff
@@ -63,6 +63,20 @@ Type getReturnTypeForMatMulOpND2D(Value A, Value B) {
...
+    Location loc, ValueRange inputs, ArrayAttr permAttr) {
+  std::vector<Value> transposedInputs;
+  for (uint64_t i = 0; i < inputs.size(); ++i) {
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/Dialect/ONNX/Rewrite.cpp]: Done. Thanks!

**Comment by tungld** [comment]
> Doing (2) would mean that pushing transpose operations to the inputs require knowledge of the local graph and, I think, this is arguably a better approach.

Yes, we can go with specific optimization to make sure that transpose fusion can be applied. I will add a constraint to check transposes on the input side soon.

**Comment by tungld** [comment]
@etiotto I added a constraint checking that all the input values are produced by TransposeOp so that transpose fusion will do it jobs to fuse multiple transposes.

**Review by etiotto — APPROVED** [review]
LGTM

**Pull request merged**